### ISSUE TITLE
Fix pod metrics in Lens for Prometheus as they display incorrect value which 2 x times bigger then actual value

### ISSUE THREAD

**Issue by dragoangel** [kontena/lens#7679]
**What would you like to be added**:
Could you add an option that will allow passing custom query parameters for metrics requests?

**Why is this needed**:
This is required for configuring aspects of monitoring, for example, pass timeout parameter to Prometheus. Also, not having the option to set query parameters when Lens is pointed to solutions like Thanos and Prometheus HA leading to displaying metrics wrongly. Lens will show duplicated data from HA, f.e.: pod CPU and RAM usage will be multiplied by the count of replicas in HA. To display data correctly and not fail on partial_response: `dedup=1&partial_response=1` would help, but Lens does not accept query parameters in `PROMETHEUS SERVICE ADDRESS` unfortunately and does not have a separate field to add them.

**Environment you are Lens application on:**

- Kubernetes distribution: Rancher
- Metrics Prometheus type: Prometheus Operator
- Metrics Prometheus service: Thanos Query Frontend that query Prometheus in HA
- Desktop OS: Any

**Comment by dragoangel** [comment]
Hm, after testing actually I found that reason of multiplied CPU and ram usage on pod compared of sum of containers not related to thanos usage and query params, because from what I tested:
1. thanos deduplicate data by default
2. on another setup without sharding and thanos issue reproduce in same way.

So this a bug. Will try debug tomorrow per query what is wrong
https://github.com/lensapp/lens/blob/2b721ab1c9073f3876758e0a345026590f957198/packages/core/src/main/prometheus/operator-provider.injectable.ts.ts#L6

**Comment by dragoangel** [comment]
I found issue at https://github.com/lensapp/lens/blob/2b721ab1c9073f3876758e0a345026590f957198/packages/core/src/main/prometheus/operator-provider.injectable.ts.ts#L71-L95

Need to add `container!=""`

@Nokel81 sorry for bothering you, but can you please check this? Thank you in advance.

**Comment by Nokel81** [comment]
I think we need to add a toggle for that, because we have added and then removed such a filter several times.

**Comment by dragoangel** [comment]
@Nokel81 you mean that previously there was such filter like `sum(kube_pod_container_resource_limits{container!="", pod=~"${opts.pods}", resource="cpu", namespace="${opts.namespace}"}) by (${opts.selector})` but it was removed due to some issue?

P.s. in general adding the option to add custom query params would be nice feature, even that not as critical as I think about it initially when was creating this issue 😊

**Comment by oleksandr-selezniov** [comment]
Hi. I'm observing doubled metric plots for pods too. And I found out that container!="" won't help for me, as my cause of duplication is two datasets with different service in Prometheus output. Those services are "kubelet" and "prometeus-kube-prometheus-kubelet"
And both datasets fit the condition container!=""

**Comment by dragoangel** [comment]
> Those services are "kubelet" and "prometeus-kube-prometheus-kubelet"

I think this issue with your jobs in prometheus that you collect same metrics twice then. You need properly setup your prometheus stack.

**Comment by vitaliyf** [comment]
Same as https://github.com/lensapp/lens/issues/7679

**Comment by Tantino** [comment]
Same as https://github.com/lensapp/lens/issues/7679

**Comment by jkroepke** [comment]
Container metrics are fine for me while pod metrics are 2x. It may depends on the installed CRI. On a AKS installed, container!="" would help.

<img width="1258" alt="image" src="https://github.com/lensapp/lens/assets/1560587/40df6cbc-c771-47f7-a977-3d34260857ad">

Running from AKS with kube-prometheus-stack.

**Comment by dragoangel** [comment]
> Container metrics are fine for me while pod metrics are 2x. It may depends on the installed CRI. On a AKS installed, container!="" would help.
> 
> <img width="1258" alt="image" src="https://github.com/lensapp/lens/assets/1560587/40df6cbc-c771-47f7-a977-3d34260857ad">
> 
> Running from AKS with kube-prometheus-stack.
> 

Yes, this exactly what I mentioned

**Comment by jkroepke** [comment]
Looking forward to #7777

**Comment by jkroepke** [comment]
I can confirm that if I run minikube with docker driver, the metric `container_memory_working_set_bytes` does not have an container label.

It depends on the CRI runtime, if the metric `container_memory_working_set_bytes` does have the container metrics or not.

@Nokel81 That could be reason why user has some trouble in history. 

If `container_memory_working_set_bytes` has a container label, then `container!=""`. If not, then `container!=""` should be not present.

#7777 would be the best solution for all.

**Comment by dragoangel** [comment]
> I can confirm that if I run minikube with docker driver, the metric `container_memory_working_set_bytes` does not have an container label.
> 
> It depends on the CRI runtime, if the metric `container_memory_working_set_bytes` does have the container metrics or not.
> 
> @Nokel81 That could be reason why user has some trouble in history.
> 
> If `container_memory_working_set_bytes` has a container label, then `container!=""`. If not, then `container!=""` should be not present.
> 
> #7777 would be the best solution for all.

If you not have container label - having promql expr with {container!=""} will not break your query...

**Comment by dragoangel** [comment]
After latest update Lens (v2023.5.310801) per container metrics broken fully 99% of time :( all of CPU\RAM\Filesystem says:
`Metrics not available at the moment`. Another problem is that downgrade Lens is not an easy thing to do...

**Comment by dragoangel** [comment]
Just for other people who struggle with latest version issues, I downgraded to 6.4.15 OpenLens to get stable monitoring tabs in Node view, Pod view, etc. All newer versions is not function properly. It anyway has issue described here, but at least other things not broken

> After latest update Lens (v2023.5.310801) per container metrics broken fully 99% of time :( all of CPU\RAM\Filesystem says: `Metrics not available at the moment`. Another problem is that downgrade Lens is not an easy thing to do...

**Comment by dragoangel** [comment]
Any updates issue?

**Comment by jkroepke** [comment]
@dragoangel I have the feeling that Lens will now develop closed sourced. May not expect anything here.

**Comment by dragoangel** [comment]
> @dragoangel I have the feeling that Lens will now develop closed sourced. May not expect anything here.

Yeah, you are totally right, as latest version of Lens is 2023.10.181418 and there are no releases here, which is bad :(. And in this version it's still same issues as was reported:

1. x2 of metrics
2. metrics for containers not available 99% of the time

**Comment by dark-brains** [comment]
Try to check your kubernetes services , I think there are services that duplicated. namespace: kube-system , services like kubelet

**Comment by SuperSecureHuman** [comment]
I belive this issue is still there.

Promethus - 

<img width="1461" height="497" alt="Image" src="https://github.com/user-attachments/assets/f76e49a2-36b6-4bfd-a01d-bc4b2e0b1389" />

Metrics Server

<img width="1453" height="594" alt="Image" src="https://github.com/user-attachments/assets/c77dca09-2bd8-438d-8100-eb8a031d0bf9" />

**Comment by SuperSecureHuman** [comment]
BTW I believe this issue should be labeled as a bug not a enhancement..

**Comment by clrothmann** [comment]
Hi @SuperSecureHuman,

Thanks for coming back with the screenshots -- but the picture on our side has actually shifted since this thread last had activity, so let me share what I found.

The container filter @dragoangel identified was added to the codebase in Jan 2025 (commit `29a0b95ca8`), behind a user-configurable toggle along the lines @Nokel81 was pointing at. All four Prometheus provider variants (lens, operator, helm, stacklight) apply `container!=""` to pod CPU and memory queries by default. There's a "Filter empty containers" checkbox in Cluster Settings → Metrics for setups where the filter breaks specific configurations.

So before we treat this as a current bug, two checks would help:

1. Which Lens version are your screenshots from? (Help → About)
2. On the cluster where you see the 2× values, is Cluster Settings → Metrics → "Filter empty containers" enabled?

One more thing on the Metrics Server screenshot -- the kubernetes metrics.k8s.io API only returns per-container rows, no pod-level aggregate, so we wouldn't expect a 2× artifact from that path. If you're genuinely seeing it there, it's most likely environmental (sidecar duplication, service mesh injection, or a Prometheus instance also feeding metrics-server). Which backend is that screenshot actually using?

If the toggle is on and you're on a recent Lens, you're sitting on something fresh and we'll want to look at it as a new investigation rather than the 2023 bug. Apologies the thread sat under "enhancement" for so long either way.

**Comment by clrothmann** [comment]
Quick follow-up -- I looked at your two screenshots more carefully. You're showing the same pod, same hour-window, ~953 MiB on the Prometheus side (kube-prometheus-stack, so operator provider in Lens) vs ~572 MiB on Metrics Server. That's a real ~1.67× discrepancy, reproducible from the data you posted -- and it's not what the Jan-2025 toggle fix would predict if both backends are wired correctly. So the questions above still matter, but for a different reason than I framed: the version + toggle status will help us narrow down whether this is a setup gap on your cluster or a code path we missed in the operator provider specifically.

**Comment by SuperSecureHuman** [comment]
Lens version - Lens: 2026.4.151333-latest

The Filter empty containers was disabled earlier.. I enabled it now, and the values match exactly the kube metrics server.. I was not aware of this setting, and should have played around more 😅 

Lemme know if you want any other details / close this issue...

**Comment by clrothmann** [comment]
Hi @SuperSecureHuman -- thanks for checking the toggle and confirming. Exact-match after enabling Filter empty containers means the operator provider is wired right; the toggle being a setting (rather than the default) is a discoverability gap on our side -- noted. Closing here; reopen if anything drifts.
